test/custom-shallow-renderer/utils.ts

```diff
@@ -0,0 +1,116 @@
+import { assert } from 'chai';
+import { options } from 'preact';
+import type { VNode } from 'preact';
+import sinon from 'sinon';
+
+// These tests are from react-shallow-renderer which uses Jest for testing, so
+// here is a lightweight adapter around `chai` to give it the same API as Jest's
+// `expect` so we don't have to rewrite all the `expect` statements to `assert`
+function createMatcher<T>(actual: T, negate = false) {
+  return {
+    toBe(expected: T) {
+      if (negate) {
+        assert.notEqual(actual, expected);
+      } else {
+        assert.equal(actual, expected);
+      }
+    },
+    toEqual(expected: T) {
+      if (negate) {
+        assert.notDeepEqual(actual, expected);
+      } else {
+        assert.deepEqual(actual, expected);
+      }
+    },
+    toBeNull() {
+      if (negate) {
+        assert.isNotNull(actual);
+      } else {
+        assert.isNull(actual);
+      }
+    },
+    toThrowError(expected: string) {
+      if (negate) {
+        assert.doesNotThrow(actual as any, expected);
+      } else {
+        assert.throws(actual as any, expected);
+      }
+    },
+    toErrorDev(expected: any, options?: any) {
+      if (negate) {
+        assert.doesNotThrow(actual as any, expected);
+      } else {
+        assert.throws(actual as any, expected);
+      }
+    },
+    toBeCalled() {
+      if (negate) {
+        sinon.assert.notCalled(actual as any);
+      } else {
+        sinon.assert.called(actual as any);
+      }
+    },
+    toHaveBeenCalled() {
+      if (negate) {
+        sinon.assert.notCalled(actual as any);
+      } else {
+        sinon.assert.called(actual as any);
+      }
+    },
+    toHaveBeenCalledTimes(count: number) {
+      if (negate) {
+        throw new Error(`.not.toHaveBeenCalledTimes is not supported`);
+      } else {
+        sinon.assert.callCount(actual as any, count);
+      }
+    },
+  };
+}
+
+type Matchers = ReturnType<typeof createMatcher>;
+type ExpectMatchers = Matchers & { not: Matchers };
+
+export function expect<T>(actual: T): ExpectMatchers {
+  const matchers: ExpectMatchers = createMatcher(actual) as any;
+  matchers.not = createMatcher(actual, true);
+
+  return matchers;
+}
+
+export function installVNodeTestHook() {
+  let prevVNodeHook: ((vnode: VNode) => void) | undefined;
+
+  before(() => {
+    prevVNodeHook = options.vnode;
+    options.vnode = vnode => {
+      if (prevVNodeHook) {
+        prevVNodeHook(vnode);
+      }
+
+      // Override the vnodeId (`__v`) to NaN so that we can compare VNodes in
+      // tests and verify expected VNode output. We choose NaN here because it
+      // successfully threads the desired behavior we want for these tests.
```

Comment:
To check my understanding here, what you're doing is disabling a Preact optimization in order to make Preact vnodes behave more like React vnodes and thereby enable a test suite originally written for React to pass?

Reply:
Yea, close. More precisely, I'm' trying to enable tests that do `expect(<div />).toBe(<div />)` which some our tests do. Our tests instead use the `stripInternalVNodeFields` to get around this `__v` issue. I went with this approach instead of using `stripInternalVNodeFields` just cuz it would be an easier change given the number of tests that compare VNodes.

It does disable a Preact optimization which is unfortunate though. The explicit goal isn't to make Preact VNodes behave like React VNodes but this change does make that more true.

Reply:
Actually, an idea a coworker had (we also have quite a few tests that do `expect(<div />`).toBe(<div />)` at work) is to make all of the internal VNodes field non-enumerable. Doing this would remove the need for the `stripInternalVNodeFields` helper altogether.

I was thinking about doing that in a separate PR but could bring it here if you wanted since we are talking about it. [Here is the relevant commit](https://github.com/preactjs/enzyme-adapter-preact-pure/commit/bedb944ad5f6bdd2ca7da2fffafac8db3279883f#diff-0ed1f7da426ac086085bde278d612a8d2d3cc18b85afa9f74f25b3f9037e5c90R5-R31).

Reply:
Okay, turns out my "make all internal fields non-enumerable" doesn't work as expected. Preact does iterate over internal VNode props to clone VNodes when rerendering so they need to be enumerable.

However, the `__v` field is specially handled so we can make just that property non-enumerable. I didn't see a place where that would break Preact but I think would enable doing things like `expect(<div />).toEqual(<div />)`. Again, I could do this in a follow up PR or bring it here. My default would be to do it in a separate PR.